language/move-lang/src/hlir/translate.rs

```diff
@@ -0,0 +1,1227 @@
+            let e = exp!(context, result, None, *te);
+            HE::Dereference(e)
+        }
+        TE::UnaryExp(op, te) => {
+            let e = exp!(context, result, None, *te);
+            HE::UnaryExp(op, e)
+        }
+        TE::BinopExp(tl, op, tr) => {
+            let el = exp!(context, result, None, *tl);
+            let er = exp!(context, result, None, *tr);
+            HE::BinopExp(el, op, er)
+        }
+        TE::Pack(_, s, tbs, tfields) => {
+            let bs = base_types(context, tbs);
+
+            let decl_fields = context.fields(&s);
+            let decl_field = |f: &Field| -> usize { *decl_fields.get(f).unwrap() };
+
+            let mut texp_fields: Vec<(usize, Field, usize, N::BaseType, T::Exp)> = tfields
+                .into_iter()
+                .map(|(f, (exp_idx, (bt, tf)))| (decl_field(&f), f, exp_idx, bt, tf))
+                .collect();
+            texp_fields.sort_by(|(_, _, eidx1, _, _), (_, _, eidx2, _, _)| eidx1.cmp(eidx2));
+
+            let mut fields = {
+                let n = decl_fields.len();
+                let mut v = Vec::with_capacity(n);
+                (0..n).for_each(|_| v.push(None));
+                v
+            };
+            for (decl_idx, f, exp_idx, bt, tf) in texp_fields {
+                let bt = base_type(context, bt);
+                let t = H::Type_::base(bt.clone());
+                let ef = exp_!(context, result, Some(&t), tf);
+                assert!(fields.get(decl_idx).unwrap().is_none());
+                if decl_idx == exp_idx {
+                    fields[decl_idx] = Some((f, bt, ef))
+                } else {
+                    let floc = ef.exp.loc;
+                    let st = H::SingleType_::base(bt.clone());
+                    let tmp = bind_exp(context, result, floc, st, ef);
+                    let move_ = HE::Move {
+                        from_user: false,
+                        var: tmp,
+                    };
+                    let move_tmp = H::exp(t, sp(eloc, move_));
+                    fields[decl_idx] = Some((f, bt, move_tmp))
+                }
+            }
+            HE::Pack(s, bs, fields.into_iter().map(|o| o.unwrap()).collect())
+        }
+        TE::ExpList(titems) => {
+            let mut items = vec![];
+            for titem in titems {
+                let item = match titem {
+                    T::ExpListItem::Single(te, ts) => {
+                        let e = exp_!(context, result, None, te);
+                        let s = single_type(context, *ts);
+                        H::ExpListItem::Single(e, Box::new(s))
+                    }
+                    T::ExpListItem::Splat(sloc, te, tss) => {
+                        let e = exp_!(context, result, None, te);
+                        let ss = single_types(context, tss);
+                        H::ExpListItem::Splat(sloc, e, ss)
+                    }
+                };
+                items.push(item)
+            }
```

Comment:
Rewrite as a map?

```rust
titems.map(|titem| {
    match titem { ... }
});

Reply:
I can't collect into an ExpResult :( See the usage of `exp_!`

Reply:
Can't you just do a `.collect<Result<_>>()?;`?

Reply:
ExpResult isn't a type alias to Result, it's its own thing.
H::ExpListItem isn't an ExpResult (nor can it be)